### PR TITLE
Integration of Promises with Threadpool

### PR DESCRIPTION
**Pull request by marcusnaslund**
Big changes, but ThreadPool add(...) behaves as before - nothing in vidproc has any objections.

### CODE REVIEW

**Comment by davidhesselbom** [comment]
I messed up and merged this even though it broke the build, hence #428. Apparently it needs some more reviewing. @emilwestergren and @simonmika, can you take a look at this?

File: source/base/ThreadPool.ooc [hunk 1]
```diff
@@ -1,111 +1,203 @@
 use ooc-collections
+use ooc-base
 
 import structs/LinkedList
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: Not used.

File: source/base/ThreadPool.ooc [hunk 2]
```diff
@@ -1,111 +1,203 @@
...
 
-SynchronizedThreadJob: class extends ThreadJob {
-	_finishedCondition := WaitCondition new()
-	_mutex := Mutex new()
-	_finished := false
-	finished ::= this _finished
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
+_Task: class {
+	_mutex: Mutex
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: `_mutex := Mutex new()`

File: source/base/ThreadPool.ooc [hunk 3]
```diff
@@ -1,111 +1,203 @@
...
+	_action: Func
+	_state : _TaskState
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: _state := _TaskState Unfinished

File: source/base/ThreadPool.ooc [hunk 4]
```diff
@@ -1,111 +1,203 @@
...
 	free: override func {
-		this _mutex lock()
-		if (!this _finished) {
-			this _freeOnCompletion = true
-			this _mutex unlock()
-		} else {
-			this _mutex unlock()
-			this _mutex destroy()
-			this _finishedCondition free()
+		if (this _state != _TaskState Unfinished)
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: What happened to `this _mutex destroy()`?

File: source/base/ThreadPool.ooc [hunk 5]
```diff
@@ -1,111 +1,203 @@
...
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
+_Task: class {
+	_mutex: Mutex
+	_action: Func
+	_state : _TaskState
+	_freeDirectly: Bool
+	init: func (=_action, =_mutex) {
+		_state = _TaskState Unfinished
+		_mutex = Mutex new()
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: `this _mutex`

File: source/base/ThreadPool.ooc [hunk 6]
```diff
@@ -1,111 +1,203 @@
...
-	finished ::= this _finished
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
+_Task: class {
+	_mutex: Mutex
+	_action: Func
+	_state : _TaskState
+	_freeDirectly: Bool
+	init: func (=_action, =_mutex) {
+		_state = _TaskState Unfinished
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: `this _state`

File: source/base/ThreadPool.ooc (hunk 5 above)
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: Both of these calls can be moved up:
```ooc
    _mutex: Mutex new()
    _state: _TaskState Unfinished
```

File: source/base/ThreadPool.ooc [hunk 7]
```diff
@@ -1,111 +1,203 @@
...
-SynchronizedThreadJob: class extends ThreadJob {
-	_finishedCondition := WaitCondition new()
-	_mutex := Mutex new()
-	_finished := false
-	finished ::= this _finished
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
+_Task: class {
+	_mutex: Mutex
+	_action: Func
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: This should not be in this class. My suggestion is to create an abstract class without any function pointers, and then you inherit two classes from it: _ActionTask and _ResultTask where the different function pointers are located.

File: source/base/ThreadPool.ooc [hunk 8]
```diff
@@ -1,111 +1,203 @@
...
+	_state : _TaskState
+	_freeDirectly: Bool
+	init: func (=_action, =_mutex) {
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: With the other changes this can be simplified to `init: func`

File: source/base/ThreadPool.ooc (hunk 5 above)
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: Wait, something's wrong here. 
```ooc
init: func (=_action, =_mutex) {
```
will assign `this _mutex` to the variable `_mutex` used as an argument in the `_Task new` call. 2 lines later, you reassign `_mutex` to `Mutex new()`. Why?

File: source/base/ThreadPool.ooc (hunk 4 above)
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: Shouldn't this be `this _state != _TaskState Finished`?

File: source/base/ThreadPool.ooc [hunk 9]
```diff
@@ -1,111 +1,203 @@
...
-			this _freeOnCompletion = true
-			this _mutex unlock()
-		} else {
-			this _mutex unlock()
-			this _mutex destroy()
-			this _finishedCondition free()
+		if (this _state != _TaskState Unfinished)
+			this _freeDirectly = true
+		else
 			super()
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: You need to free the Mutex before you call super.

File: source/base/ThreadPool.ooc [hunk 10]
```diff
@@ -1,111 +1,203 @@
...
-		}
 	}
-	run: override func {
-		this _execute()
-		this _mutex lock()
-		this _finished = true
-		if (this _freeOnCompletion) {
-			this _mutex unlock()
+	run: virtual func (=_mutex) {
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: This again reassigns `this _mutex` without destroying the current one. Why?

File: source/base/ThreadPool.ooc [hunk 11]
```diff
@@ -1,111 +1,203 @@
...
-	_mutex := Mutex new()
-	_finished := false
-	finished ::= this _finished
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
+_Task: class {
+	_mutex: Mutex
+	_action: Func
+	_state : _TaskState
+	_freeDirectly: Bool
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: `_freeDirectly := false`

File: source/base/ThreadPool.ooc [hunk 12]
```diff
@@ -1,111 +1,203 @@
...
+	init: func (=_action, =_mutex) {
+		_state = _TaskState Unfinished
+		_mutex = Mutex new()
+	}
 	free: override func {
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: This function needs to be thread-safe, use the mutex.

File: source/base/ThreadPool.ooc (hunk 10 above)
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: Change this to `run: abstract func` and implement in concrete classes.

File: source/base/ThreadPool.ooc [hunk 13]
```diff
@@ -1,111 +1,203 @@
...
 			this _threads[i] start()
 		}
 	}
 	free: override func {
-		for (i in 0 .. this _threadCount)
+		for (i in 0 .. this _threadCount) {
 			this _threads[i] free()
+			this _threadMutexes[i] destroy()
+		}
 		gc_free(this _threads data)
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: I think `this _threads free()` should work, too.

File: source/base/ThreadPool.ooc [hunk 14]
```diff
@@ -1,111 +1,203 @@
...
 	}
-	wait: func {
+}
+
+_ResultTask: class <T> extends _Task {
+	_result: Cell<T>
+	_resultAction: Func -> T
+	init: func (=_resultAction, =_mutex) {
+		this _state = _TaskState Unfinished
+	}
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: With the other changes, this can be simplified to `init: func(=_resultAction) { super() }`

File: source/base/ThreadPool.ooc (hunk 13 above)
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: I don't see `_this threadMutexes` getting freed anywhere...

File: source/base/ThreadPool.ooc [hunk 15]
```diff
@@ -1,111 +1,203 @@
...
+	free: func {
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: This function is not needed since it is already implemented in the base class.

File: source/base/ThreadPool.ooc [hunk 16]
```diff
@@ -1,111 +1,203 @@
...
+		if (this _state != _TaskState Unfinished)
+			this _freeDirectly = true
+		else
+			super()
+	}
+	run: override func (=_mutex) {
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: this should be `run: override func {`

File: source/base/ThreadPool.ooc [hunk 17]
```diff
@@ -1,111 +1,203 @@
...
-			this _mutex lock()
+			job := this _tasks wait()
+			job run(this _threadMutexes[threadId])
+			this _globalmutex lock()
 			this _activeJobs -= 1
-			if (this _activeJobs == 0)
+			if (this _activeJobs == 0) {
 				this _allFinishedCondition broadcast()
-			this _mutex unlock()
+			}
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: These brackets are unnecessary.

File: source/base/ThreadPool.ooc [hunk 18]
```diff
@@ -1,111 +1,203 @@
...
 }
 
+_TaskPromise: class extends Promise {
+	_task: _Task
+	init: func (=_task)
+	free: override func {
+		this _task free()
+		super()
+	}
+	wait: override func -> Bool {
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: This is not a very efficient way of synchronizing. Use Conditions like in the earlier ThreadPool implementation.

File: source/base/ThreadPool.ooc [hunk 19]
```diff
@@ -1,111 +1,203 @@
...
+	_add: func (task: _Task) -> Void {
+		this _globalmutex lock()
 		this _activeJobs += 1
-		this _mutex unlock()
-		this _jobs enqueue(job)
+		this _globalmutex unlock()
+		this _tasks enqueue(task)
+	}
+	add: func (action: Func) {
+		task := _Task new(action, this _globalmutex)
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: Okay, so... tasks are given the `globalmutex`...? So that's why they're never freed in `_Task`.

File: source/base/ThreadPool.ooc [hunk 20]
```diff
@@ -1,111 +1,203 @@
...
+			this _resultTask _state = _TaskState Cancelled
+			status = true
+		}
+		this _resultTask _mutex unlock()
+		status
+	}
+}
+
 ThreadPool: class {
+	_globalmutex := Mutex new()
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: `_globalMutex`

File: source/base/ThreadPool.ooc [hunk 21]
```diff
@@ -1,111 +1,203 @@
...
+				status = (this _task _state == _TaskState Finished)
+				this _task _mutex unlock()
+				break
+			}
+			this _task _mutex unlock()
+			Time sleepMilli(10)
+		}
+		status
+	}
+	cancel: override func -> Bool {
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: This currently has no effect on the Task, is this something for the future?

File: source/base/ThreadPool.ooc [hunk 22]
```diff
@@ -1,111 +1,203 @@
...
 	}
-	run: override func {
-		this _execute()
-		this _mutex lock()
-		this _finished = true
-		if (this _freeOnCompletion) {
-			this _mutex unlock()
+	run: virtual func (=_mutex) {
+		this _action()
+		_mutex lock()
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: `this _mutex`

File: source/base/ThreadPool.ooc (hunk 5 above)
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: Okay, I'm starting to see what's going on here. Still, if you create a new `Mutex` here in the constructor and don't destroy it later, it will leak.

File: source/base/ThreadPool.ooc (hunk 19 above)
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: Why not have private mutexes for the tasks?

File: source/base/ThreadPool.ooc (hunk 5 above)
Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: This was very unfortunate. There should not be a new mutex created here.

File: source/base/ThreadPool.ooc (hunk 18 above)
Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: Since we have the new (in the future platform independent?) WaitCondition class, absolutely.

File: source/base/ThreadPool.ooc [hunk 23]
```diff
@@ -1,111 +1,203 @@
...
-		job := SynchronizedThreadJob new(body)
-		this _add(job)
-		job
+	addWaitResult: func ~result <T> (action: Func -> T) -> ResultPromise<T> {
+		task := _ResultTask<T> new(action, this _globalmutex)
+		this _add(task)
+		_ResultTaskPromise new(task)
 	}
-	add: func (body: Func) { this _add(ThreadJob new(body)) }
-	wait: func (body: Func) { this addSynchronized(body) wait() . free() }
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: Why did you remove this? It's nice to have when you want to add something to the pool and synchronize immediately after, without having to deal with waiting on and freeing promises.

File: source/base/ThreadPool.ooc (hunk 19 above)
Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: To avoid the cost of creating a new mutex for each task, since there may be many.

Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: Don't worry about that, we can solve this in the long run by recycling Mutexes.

File: source/base/ThreadPool.ooc (hunk 23 above)
Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: Good point, I removed it because it was related to the now removed SynchronizedThreadJob. I will make it work again.

File: source/base/ThreadPool.ooc (hunk 3 above)
Comment (simonmika) [inline source/base/ThreadPool.ooc]: That is the default value. There is no difference.

File: source/base/ThreadPool.ooc (hunk 11 above)
Comment (simonmika) [inline source/base/ThreadPool.ooc]: Again that is the default value already, no need for unnecessary assignments.

File: source/base/ThreadPool.ooc (hunk 19 above)
Comment (simonmika) [inline source/base/ThreadPool.ooc]: It is enough to have a mutex per thread so why ever create more? Even if you recycle.

File: source/base/ThreadPool.ooc (hunk 23 above)
Comment (simonmika) [inline source/base/ThreadPool.ooc]: I don't think that it should be there. Wait for the task you added.

File: source/base/ThreadPool.ooc (hunk 11 above)
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: We've had trouble with uninitialized values before, `value: Int` not being 0 before being assigned etc. On the other hand, the case might be different with data type members of classes, since Objects are instantiated with a `malloc` call...

File: source/base/ThreadPool.ooc (hunk 19 above)
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: It causes unnecessary waiting. Why share a mutex when they are independent of eachother?

File: source/base/ThreadPool.ooc (hunk 21 above)
Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: Ideally, this would interrupt/cancel the thread and have it move on to the next task in the queue (if any). Haven't yet figured out a nice way to do this though.

Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: For now, it allows wait() to return directly (although the task will still run).

File: source/base/ThreadPool.ooc (hunk 23 above)
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: If you say so. Just remember to adapt the Android code in ooc-vidproc because this breaks the kean API.
